{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Install NCBI Datasets command line tool and API library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!curl -o datasets 'https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets' && chmod +x datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip3 install ncbi-datasets-pylib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip3 install biopython\n",
    "!pip3 install pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import zipfile\n",
    "import pandas as pd\n",
    "import yaml\n",
    "from pprint import pprint\n",
    "from datetime import datetime\n",
    "from collections import defaultdict, Counter\n",
    "from IPython.display import display\n",
    "from Bio.SeqFeature import FeatureLocation, CompoundLocation\n",
    "\n",
    "try:\n",
    "    import ncbi.datasets\n",
    "except ImportError:\n",
    "    print('ncbi.datasets module not found. To install, run `pip install ncbi-datasets-pylib`.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Documentation of some useful APIs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class FeatureLocation in module Bio.SeqFeature:\n",
      "\n",
      "class FeatureLocation(builtins.object)\n",
      " |  FeatureLocation(start, end, strand=None, ref=None, ref_db=None)\n",
      " |  \n",
      " |  Specify the location of a feature along a sequence.\n",
      " |  \n",
      " |  The FeatureLocation is used for simple continuous features, which can\n",
      " |  be described as running from a start position to and end position\n",
      " |  (optionally with a strand and reference information).  More complex\n",
      " |  locations made up from several non-continuous parts (e.g. a coding\n",
      " |  sequence made up of several exons) are described using a SeqFeature\n",
      " |  with a CompoundLocation.\n",
      " |  \n",
      " |  Note that the start and end location numbering follow Python's scheme,\n",
      " |  thus a GenBank entry of 123..150 (one based counting) becomes a location\n",
      " |  of [122:150] (zero based counting).\n",
      " |  \n",
      " |  >>> from Bio.SeqFeature import FeatureLocation\n",
      " |  >>> f = FeatureLocation(122, 150)\n",
      " |  >>> print(f)\n",
      " |  [122:150]\n",
      " |  >>> print(f.start)\n",
      " |  122\n",
      " |  >>> print(f.end)\n",
      " |  150\n",
      " |  >>> print(f.strand)\n",
      " |  None\n",
      " |  \n",
      " |  Note the strand defaults to None. If you are working with nucleotide\n",
      " |  sequences you'd want to be explicit if it is the forward strand:\n",
      " |  \n",
      " |  >>> from Bio.SeqFeature import FeatureLocation\n",
      " |  >>> f = FeatureLocation(122, 150, strand=+1)\n",
      " |  >>> print(f)\n",
      " |  [122:150](+)\n",
      " |  >>> print(f.strand)\n",
      " |  1\n",
      " |  \n",
      " |  Note that for a parent sequence of length n, the FeatureLocation\n",
      " |  start and end must satisfy the inequality 0 <= start <= end <= n.\n",
      " |  This means even for features on the reverse strand of a nucleotide\n",
      " |  sequence, we expect the 'start' coordinate to be less than the\n",
      " |  'end'.\n",
      " |  \n",
      " |  >>> from Bio.SeqFeature import FeatureLocation\n",
      " |  >>> r = FeatureLocation(122, 150, strand=-1)\n",
      " |  >>> print(r)\n",
      " |  [122:150](-)\n",
      " |  >>> print(r.start)\n",
      " |  122\n",
      " |  >>> print(r.end)\n",
      " |  150\n",
      " |  >>> print(r.strand)\n",
      " |  -1\n",
      " |  \n",
      " |  i.e. Rather than thinking of the 'start' and 'end' biologically in a\n",
      " |  strand aware manner, think of them as the 'left most' or 'minimum'\n",
      " |  boundary, and the 'right most' or 'maximum' boundary of the region\n",
      " |  being described. This is particularly important with compound\n",
      " |  locations describing non-continuous regions.\n",
      " |  \n",
      " |  In the example above we have used standard exact positions, but there\n",
      " |  are also specialised position objects used to represent fuzzy positions\n",
      " |  as well, for example a GenBank location like complement(<123..150)\n",
      " |  would use a BeforePosition object for the start.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Combine location with another FeatureLocation object, or shift it.\n",
      " |      \n",
      " |      You can add two feature locations to make a join CompoundLocation:\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation\n",
      " |      >>> f1 = FeatureLocation(5, 10)\n",
      " |      >>> f2 = FeatureLocation(20, 30)\n",
      " |      >>> combined = f1 + f2\n",
      " |      >>> print(combined)\n",
      " |      join{[5:10], [20:30]}\n",
      " |      \n",
      " |      This is thus equivalent to:\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import CompoundLocation\n",
      " |      >>> join = CompoundLocation([f1, f2])\n",
      " |      >>> print(join)\n",
      " |      join{[5:10], [20:30]}\n",
      " |      \n",
      " |      You can also use sum(...) in this way:\n",
      " |      \n",
      " |      >>> join = sum([f1, f2])\n",
      " |      >>> print(join)\n",
      " |      join{[5:10], [20:30]}\n",
      " |      \n",
      " |      Furthermore, you can combine a FeatureLocation with a CompoundLocation\n",
      " |      in this way.\n",
      " |      \n",
      " |      Separately, adding an integer will give a new FeatureLocation with\n",
      " |      its start and end offset by that amount. For example:\n",
      " |      \n",
      " |      >>> print(f1)\n",
      " |      [5:10]\n",
      " |      >>> print(f1 + 100)\n",
      " |      [105:110]\n",
      " |      >>> print(200 + f1)\n",
      " |      [205:210]\n",
      " |      \n",
      " |      This can be useful when editing annotation.\n",
      " |  \n",
      " |  __contains__(self, value)\n",
      " |      Check if an integer position is within the FeatureLocation object.\n",
      " |      \n",
      " |      Note that extra care may be needed for fuzzy locations, e.g.\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation\n",
      " |      >>> from Bio.SeqFeature import BeforePosition, AfterPosition\n",
      " |      >>> loc = FeatureLocation(BeforePosition(5), AfterPosition(10))\n",
      " |      >>> len(loc)\n",
      " |      5\n",
      " |      >>> [i for i in range(15) if i in loc]\n",
      " |      [5, 6, 7, 8, 9]\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Implement equality by comparing all the location attributes.\n",
      " |  \n",
      " |  __init__(self, start, end, strand=None, ref=None, ref_db=None)\n",
      " |      Initialize the class.\n",
      " |      \n",
      " |      start and end arguments specify the values where the feature begins\n",
      " |      and ends. These can either by any of the ``*Position`` objects that\n",
      " |      inherit from AbstractPosition, or can just be integers specifying the\n",
      " |      position. In the case of integers, the values are assumed to be\n",
      " |      exact and are converted in ExactPosition arguments. This is meant\n",
      " |      to make it easy to deal with non-fuzzy ends.\n",
      " |      \n",
      " |      i.e. Short form:\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation\n",
      " |      >>> loc = FeatureLocation(5, 10, strand=-1)\n",
      " |      >>> print(loc)\n",
      " |      [5:10](-)\n",
      " |      \n",
      " |      Explicit form:\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation, ExactPosition\n",
      " |      >>> loc = FeatureLocation(ExactPosition(5), ExactPosition(10), strand=-1)\n",
      " |      >>> print(loc)\n",
      " |      [5:10](-)\n",
      " |      \n",
      " |      Other fuzzy positions are used similarly,\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation\n",
      " |      >>> from Bio.SeqFeature import BeforePosition, AfterPosition\n",
      " |      >>> loc2 = FeatureLocation(BeforePosition(5), AfterPosition(10), strand=-1)\n",
      " |      >>> print(loc2)\n",
      " |      [<5:>10](-)\n",
      " |      \n",
      " |      For nucleotide features you will also want to specify the strand,\n",
      " |      use 1 for the forward (plus) strand, -1 for the reverse (negative)\n",
      " |      strand, 0 for stranded but strand unknown (? in GFF3), or None for\n",
      " |      when the strand does not apply (dot in GFF3), e.g. features on\n",
      " |      proteins.\n",
      " |      \n",
      " |      >>> loc = FeatureLocation(5, 10, strand=+1)\n",
      " |      >>> print(loc)\n",
      " |      [5:10](+)\n",
      " |      >>> print(loc.strand)\n",
      " |      1\n",
      " |      \n",
      " |      Normally feature locations are given relative to the parent\n",
      " |      sequence you are working with, but an explicit accession can\n",
      " |      be given with the optional ref and db_ref strings:\n",
      " |      \n",
      " |      >>> loc = FeatureLocation(105172, 108462, ref=\"AL391218.9\", strand=1)\n",
      " |      >>> print(loc)\n",
      " |      AL391218.9[105172:108462](+)\n",
      " |      >>> print(loc.ref)\n",
      " |      AL391218.9\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over the parent positions within the FeatureLocation object.\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation\n",
      " |      >>> from Bio.SeqFeature import BeforePosition, AfterPosition\n",
      " |      >>> loc = FeatureLocation(BeforePosition(5), AfterPosition(10))\n",
      " |      >>> len(loc)\n",
      " |      5\n",
      " |      >>> for i in loc: print(i)\n",
      " |      5\n",
      " |      6\n",
      " |      7\n",
      " |      8\n",
      " |      9\n",
      " |      >>> list(loc)\n",
      " |      [5, 6, 7, 8, 9]\n",
      " |      >>> [i for i in range(15) if i in loc]\n",
      " |      [5, 6, 7, 8, 9]\n",
      " |      \n",
      " |      Note this is strand aware:\n",
      " |      \n",
      " |      >>> loc = FeatureLocation(BeforePosition(5), AfterPosition(10), strand = -1)\n",
      " |      >>> list(loc)\n",
      " |      [9, 8, 7, 6, 5]\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the region described by the FeatureLocation object.\n",
      " |      \n",
      " |      Note that extra care may be needed for fuzzy locations, e.g.\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation\n",
      " |      >>> from Bio.SeqFeature import BeforePosition, AfterPosition\n",
      " |      >>> loc = FeatureLocation(BeforePosition(5), AfterPosition(10))\n",
      " |      >>> len(loc)\n",
      " |      5\n",
      " |  \n",
      " |  __nonzero__(self)\n",
      " |      Return True regardless of the length of the feature.\n",
      " |      \n",
      " |      This behaviour is for backwards compatibility, since until the\n",
      " |      __len__ method was added, a FeatureLocation always evaluated as True.\n",
      " |      \n",
      " |      Note that in comparison, Seq objects, strings, lists, etc, will all\n",
      " |      evaluate to False if they have length zero.\n",
      " |      \n",
      " |      WARNING: The FeatureLocation may in future evaluate to False when its\n",
      " |      length is zero (in order to better match normal python behaviour)!\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add a feature locationanother FeatureLocation object to the left.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Represent the FeatureLocation object as a string for debugging.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return a representation of the FeatureLocation object (with python counting).\n",
      " |      \n",
      " |      For the simple case this uses the python splicing syntax, [122:150]\n",
      " |      (zero based counting) which GenBank would call 123..150 (one based\n",
      " |      counting).\n",
      " |  \n",
      " |  extract(self, parent_sequence)\n",
      " |      Extract the sequence from supplied parent sequence using the FeatureLocation object.\n",
      " |      \n",
      " |      The parent_sequence can be a Seq like object or a string, and will\n",
      " |      generally return an object of the same type. The exception to this is\n",
      " |      a MutableSeq as the parent sequence will return a Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_protein\n",
      " |      >>> from Bio.SeqFeature import FeatureLocation\n",
      " |      >>> seq = Seq(\"MKQHKAMIVALIVICITAVVAAL\", generic_protein)\n",
      " |      >>> feature_loc = FeatureLocation(8, 15)\n",
      " |      >>> feature_loc.extract(seq)\n",
      " |      Seq('VALIVIC', ProteinAlphabet())\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  end\n",
      " |      End location - right most (maximum) value, regardless of strand.\n",
      " |      \n",
      " |      Read only, returns an integer like position object, possibly a fuzzy\n",
      " |      position.\n",
      " |  \n",
      " |  nofuzzy_end\n",
      " |      End position (integer, approximated if fuzzy, read only) (OBSOLETE).\n",
      " |      \n",
      " |      This is now an alias for int(feature.end), which should be\n",
      " |      used in preference -- unless you are trying to support old\n",
      " |      versions of Biopython.\n",
      " |  \n",
      " |  nofuzzy_start\n",
      " |      Start position (integer, approximated if fuzzy, read only) (OBSOLETE).\n",
      " |      \n",
      " |      This is now an alias for int(feature.start), which should be\n",
      " |      used in preference -- unless you are trying to support old\n",
      " |      versions of Biopython.\n",
      " |  \n",
      " |  parts\n",
      " |      Read only list of sections (always one, the FeatureLocation object).\n",
      " |      \n",
      " |      This is a convenience property allowing you to write code handling\n",
      " |      both simple FeatureLocation objects (with one part) and more complex\n",
      " |      CompoundLocation objects (with multiple parts) interchangeably.\n",
      " |  \n",
      " |  start\n",
      " |      Start location - left most (minimum) value, regardless of strand.\n",
      " |      \n",
      " |      Read only, returns an integer like position object, possibly a fuzzy\n",
      " |      position.\n",
      " |  \n",
      " |  strand\n",
      " |      Strand of the location (+1, -1, 0 or None).\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __hash__ = None\n",
      "\n",
      "Help on class CompoundLocation in module Bio.SeqFeature:\n",
      "\n",
      "class CompoundLocation(builtins.object)\n",
      " |  CompoundLocation(parts, operator='join')\n",
      " |  \n",
      " |  For handling joins etc where a feature location has several parts.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Combine locations, or shift the location by an integer offset.\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation, CompoundLocation\n",
      " |      >>> f1 = FeatureLocation(15, 17) + FeatureLocation(20, 30)\n",
      " |      >>> print(f1)\n",
      " |      join{[15:17], [20:30]}\n",
      " |      \n",
      " |      You can add another FeatureLocation:\n",
      " |      \n",
      " |      >>> print(f1 + FeatureLocation(40, 50))\n",
      " |      join{[15:17], [20:30], [40:50]}\n",
      " |      >>> print(FeatureLocation(5, 10) + f1)\n",
      " |      join{[5:10], [15:17], [20:30]}\n",
      " |      \n",
      " |      You can also add another CompoundLocation:\n",
      " |      \n",
      " |      >>> f2 = FeatureLocation(40, 50) + FeatureLocation(60, 70)\n",
      " |      >>> print(f2)\n",
      " |      join{[40:50], [60:70]}\n",
      " |      >>> print(f1 + f2)\n",
      " |      join{[15:17], [20:30], [40:50], [60:70]}\n",
      " |      \n",
      " |      Also, as with the FeatureLocation, adding an integer shifts the\n",
      " |      location's co-ordinates by that offset:\n",
      " |      \n",
      " |      >>> print(f1 + 100)\n",
      " |      join{[115:117], [120:130]}\n",
      " |      >>> print(200 + f1)\n",
      " |      join{[215:217], [220:230]}\n",
      " |      >>> print(f1 + (-5))\n",
      " |      join{[10:12], [15:25]}\n",
      " |  \n",
      " |  __contains__(self, value)\n",
      " |      Check if an integer position is within the CompoundLocation object.\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Check if all parts of CompoundLocation are equal to all parts of other CompoundLocation.\n",
      " |  \n",
      " |  __init__(self, parts, operator='join')\n",
      " |      Initialize the class.\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation, CompoundLocation\n",
      " |      >>> f1 = FeatureLocation(10, 40, strand=+1)\n",
      " |      >>> f2 = FeatureLocation(50, 59, strand=+1)\n",
      " |      >>> f = CompoundLocation([f1, f2])\n",
      " |      >>> len(f) == len(f1) + len(f2) == 39 == len(list(f))\n",
      " |      True\n",
      " |      >>> print(f.operator)\n",
      " |      join\n",
      " |      >>> 5 in f\n",
      " |      False\n",
      " |      >>> 15 in f\n",
      " |      True\n",
      " |      >>> f.strand\n",
      " |      1\n",
      " |      \n",
      " |      Notice that the strand of the compound location is computed\n",
      " |      automatically - in the case of mixed strands on the sub-locations\n",
      " |      the overall strand is set to None.\n",
      " |      \n",
      " |      >>> f = CompoundLocation([FeatureLocation(3, 6, strand=+1),\n",
      " |      ...                       FeatureLocation(10, 13, strand=-1)])\n",
      " |      >>> print(f.strand)\n",
      " |      None\n",
      " |      >>> len(f)\n",
      " |      6\n",
      " |      >>> list(f)\n",
      " |      [3, 4, 5, 12, 11, 10]\n",
      " |      \n",
      " |      The example above doing list(f) iterates over the coordinates within the\n",
      " |      feature. This allows you to use max and min on the location, to find the\n",
      " |      range covered:\n",
      " |      \n",
      " |      >>> min(f)\n",
      " |      3\n",
      " |      >>> max(f)\n",
      " |      12\n",
      " |      \n",
      " |      More generally, you can use the compound location's start and end which\n",
      " |      give the full span covered, 0 <= start <= end <= full sequence length.\n",
      " |      \n",
      " |      >>> f.start == min(f)\n",
      " |      True\n",
      " |      >>> f.end == max(f) + 1\n",
      " |      True\n",
      " |      \n",
      " |      This is consistent with the behaviour of the simple FeatureLocation for\n",
      " |      a single region, where again the 'start' and 'end' do not necessarily\n",
      " |      give the biological start and end, but rather the 'minimal' and 'maximal'\n",
      " |      coordinate boundaries.\n",
      " |      \n",
      " |      Note that adding locations provides a more intuitive method of\n",
      " |      construction:\n",
      " |      \n",
      " |      >>> f = FeatureLocation(3, 6, strand=+1) + FeatureLocation(10, 13, strand=-1)\n",
      " |      >>> len(f)\n",
      " |      6\n",
      " |      >>> list(f)\n",
      " |      [3, 4, 5, 12, 11, 10]\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over the parent positions within the CompoundLocation object.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the CompoundLocation object.\n",
      " |  \n",
      " |  __nonzero__(self)\n",
      " |      Return True regardless of the length of the feature.\n",
      " |      \n",
      " |      This behaviour is for backwards compatibility, since until the\n",
      " |      __len__ method was added, a FeatureLocation always evaluated as True.\n",
      " |      \n",
      " |      Note that in comparison, Seq objects, strings, lists, etc, will all\n",
      " |      evaluate to False if they have length zero.\n",
      " |      \n",
      " |      WARNING: The FeatureLocation may in future evaluate to False when its\n",
      " |      length is zero (in order to better match normal python behaviour)!\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add a feature to the left.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Represent the CompoundLocation object as string for debugging.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return a representation of the CompoundLocation object (with python counting).\n",
      " |  \n",
      " |  extract(self, parent_sequence)\n",
      " |      Extract the sequence from supplied parent sequence using the CompoundLocation object.\n",
      " |      \n",
      " |      The parent_sequence can be a Seq like object or a string, and will\n",
      " |      generally return an object of the same type. The exception to this is\n",
      " |      a MutableSeq as the parent sequence will return a Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_protein\n",
      " |      >>> from Bio.SeqFeature import FeatureLocation, CompoundLocation\n",
      " |      >>> seq = Seq(\"MKQHKAMIVALIVICITAVVAAL\", generic_protein)\n",
      " |      >>> fl1 = FeatureLocation(2, 8)\n",
      " |      >>> fl2 = FeatureLocation(10, 15)\n",
      " |      >>> fl3 = CompoundLocation([fl1,fl2])\n",
      " |      >>> fl3.extract(seq)\n",
      " |      Seq('QHKAMILIVIC', ProteinAlphabet())\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  end\n",
      " |      End location - right most (maximum) value, regardless of strand.\n",
      " |      \n",
      " |      Read only, returns an integer like position object, possibly a fuzzy\n",
      " |      position.\n",
      " |      \n",
      " |      For the special case of a CompoundLocation wrapping the origin of\n",
      " |      a circular genome this will match the genome length (minus one\n",
      " |      given how Python counts from zero).\n",
      " |  \n",
      " |  nofuzzy_end\n",
      " |      End position (integer, approximated if fuzzy, read only) (OBSOLETE).\n",
      " |      \n",
      " |      This is an alias for int(feature.end), which should be used in\n",
      " |      preference -- unless you are trying to support old versions of\n",
      " |      Biopython.\n",
      " |  \n",
      " |  nofuzzy_start\n",
      " |      Start position (integer, approximated if fuzzy, read only) (OBSOLETE).\n",
      " |      \n",
      " |      This is an alias for int(feature.start), which should be used in\n",
      " |      preference -- unless you are trying to support old versions of\n",
      " |      Biopython.\n",
      " |  \n",
      " |  ref\n",
      " |      Not present in CompoundLocation, dummy method for API compatibility.\n",
      " |  \n",
      " |  ref_db\n",
      " |      Not present in CompoundLocation, dummy method for API compatibility.\n",
      " |  \n",
      " |  start\n",
      " |      Start location - left most (minimum) value, regardless of strand.\n",
      " |      \n",
      " |      Read only, returns an integer like position object, possibly a fuzzy\n",
      " |      position.\n",
      " |      \n",
      " |      For the special case of a CompoundLocation wrapping the origin of a\n",
      " |      circular genome, this will return zero.\n",
      " |  \n",
      " |  strand\n",
      " |      Overall strand of the compound location.\n",
      " |      \n",
      " |      If all the parts have the same strand, that is returned. Otherwise\n",
      " |      for mixed strands, this returns None.\n",
      " |      \n",
      " |      >>> from Bio.SeqFeature import FeatureLocation, CompoundLocation\n",
      " |      >>> f1 = FeatureLocation(15, 17, strand=1)\n",
      " |      >>> f2 = FeatureLocation(20, 30, strand=-1)\n",
      " |      >>> f = f1 + f2\n",
      " |      >>> f1.strand\n",
      " |      1\n",
      " |      >>> f2.strand\n",
      " |      -1\n",
      " |      >>> f.strand\n",
      " |      >>> f.strand is None\n",
      " |      True\n",
      " |      \n",
      " |      If you set the strand of a CompoundLocation, this is applied to\n",
      " |      all the parts - use with caution:\n",
      " |      \n",
      " |      >>> f.strand = 1\n",
      " |      >>> f1.strand\n",
      " |      1\n",
      " |      >>> f2.strand\n",
      " |      1\n",
      " |      >>> f.strand\n",
      " |      1\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __hash__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(Bio.SeqFeature.FeatureLocation)\n",
    "help(Bio.SeqFeature.CompoundLocation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NOTE: Failed attempt to use NCBI Datasets API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with ncbi.datasets.ApiClient() as api_client:\n",
    "    # Create an instance of the API class\n",
    "    api_instance = ncbi.datasets.DownloadApi(api_client)\n",
    "    gene_ids = [7113]\n",
    "    include_sequence_type = ['SEQ_TYPE_GENE', 'SEQ_TYPE_RNA', 'SEQ_TYPE_PROTEIN']\n",
    "    filename = 'sars-cov2-genes'\n",
    "    \n",
    "    try:\n",
    "        api_response = api_instance.download_gene_package(gene_ids,\n",
    "                            filename=filename)\n",
    "    except ncbi.datasets.ApiException as e:\n",
    "        print(\"Exception when calling DownloadApi->download_gene_package: %s\\n\" % e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ALTERNATIVE: Use NCBI Datasets command line tool."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!./datasets download gene 59272"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! rm -rf sars-cov2-gene-data\n",
    "! unzip -d sars-cov2-gene-data ncbi_dataset.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'genes': [{'chromosomes': ['X'],\n",
      "            'commonName': 'human',\n",
      "            'description': 'angiotensin I converting enzyme 2',\n",
      "            'ensemblGeneIds': ['ENSG00000130234'],\n",
      "            'geneId': '59272',\n",
      "            'genomicRanges': [{'accessionVersion': 'NC_000023.11',\n",
      "                               'range': [{'begin': '15561033',\n",
      "                                          'end': '15602158',\n",
      "                                          'orientation': 'minus'}]}],\n",
      "            'nomenclatureAuthority': {'authority': 'HGNC',\n",
      "                                      'identifier': 'HGNC:13557'},\n",
      "            'omimIds': ['300335'],\n",
      "            'swissProtAccessions': ['Q9BYF1.2'],\n",
      "            'symbol': 'ACE2',\n",
      "            'synonyms': ['ACEH'],\n",
      "            'taxId': '9606',\n",
      "            'taxname': 'Homo sapiens',\n",
      "            'transcripts': [{'accessionVersion': 'NM_021804.3',\n",
      "                             'cds': {'accessionVersion': 'NM_021804.3',\n",
      "                                     'range': [{'begin': '307',\n",
      "                                                'end': '2724'}]},\n",
      "                             'ensemblTranscript': 'ENST00000427411.1',\n",
      "                             'exons': {'accessionVersion': 'NC_000023.11',\n",
      "                                       'range': [{'begin': '15601956',\n",
      "                                                  'end': '15602158',\n",
      "                                                  'order': 1},\n",
      "                                                 {'begin': '15600726',\n",
      "                                                  'end': '15601014',\n",
      "                                                  'order': 2},\n",
      "                                                 {'begin': '15594845',\n",
      "                                                  'end': '15595003',\n",
      "                                                  'order': 3},\n",
      "                                                 {'begin': '15592229',\n",
      "                                                  'end': '15592322',\n",
      "                                                  'order': 4},\n",
      "                                                 {'begin': '15591713',\n",
      "                                                  'end': '15591856',\n",
      "                                                  'order': 5},\n",
      "                                                 {'begin': '15589344',\n",
      "                                                  'end': '15589456',\n",
      "                                                  'order': 6},\n",
      "                                                 {'begin': '15587753',\n",
      "                                                  'end': '15587858',\n",
      "                                                  'order': 7},\n",
      "                                                 {'begin': '15585475',\n",
      "                                                  'end': '15585572',\n",
      "                                                  'order': 8},\n",
      "                                                 {'begin': '15581221',\n",
      "                                                  'end': '15581390',\n",
      "                                                  'order': 9},\n",
      "                                                 {'begin': '15578089',\n",
      "                                                  'end': '15578315',\n",
      "                                                  'order': 10},\n",
      "                                                 {'begin': '15575666',\n",
      "                                                  'end': '15575810',\n",
      "                                                  'order': 11},\n",
      "                                                 {'begin': '15573367',\n",
      "                                                  'end': '15573465',\n",
      "                                                  'order': 12},\n",
      "                                                 {'begin': '15572201',\n",
      "                                                  'end': '15572323',\n",
      "                                                  'order': 13},\n",
      "                                                 {'begin': '15571624',\n",
      "                                                  'end': '15571796',\n",
      "                                                  'order': 14},\n",
      "                                                 {'begin': '15570295',\n",
      "                                                  'end': '15570353',\n",
      "                                                  'order': 15},\n",
      "                                                 {'begin': '15567726',\n",
      "                                                  'end': '15567826',\n",
      "                                                  'order': 16},\n",
      "                                                 {'begin': '15566253',\n",
      "                                                  'end': '15566369',\n",
      "                                                  'order': 17},\n",
      "                                                 {'begin': '15564024',\n",
      "                                                  'end': '15564218',\n",
      "                                                  'order': 18},\n",
      "                                                 {'begin': '15561033',\n",
      "                                                  'end': '15562013',\n",
      "                                                  'order': 19}]},\n",
      "                             'length': 3596,\n",
      "                             'name': 'transcript variant 2',\n",
      "                             'protein': {'accessionVersion': 'NP_068576.1',\n",
      "                                         'ensemblProtein': 'ENSP00000389326.1',\n",
      "                                         'length': 805,\n",
      "                                         'name': 'angiotensin-converting '\n",
      "                                                 'enzyme 2 precursor'},\n",
      "                             'type': 'PROTEIN_CODING'},\n",
      "                            {'accessionVersion': 'NM_001371415.1',\n",
      "                             'cds': {'accessionVersion': 'NM_001371415.1',\n",
      "                                     'range': [{'begin': '50', 'end': '2467'}]},\n",
      "                             'ensemblTranscript': 'ENST00000252519.8',\n",
      "                             'exons': {'accessionVersion': 'NC_000023.11',\n",
      "                                       'range': [{'begin': '15600726',\n",
      "                                                  'end': '15600960',\n",
      "                                                  'order': 1},\n",
      "                                                 {'begin': '15594845',\n",
      "                                                  'end': '15595003',\n",
      "                                                  'order': 2},\n",
      "                                                 {'begin': '15592229',\n",
      "                                                  'end': '15592322',\n",
      "                                                  'order': 3},\n",
      "                                                 {'begin': '15591713',\n",
      "                                                  'end': '15591856',\n",
      "                                                  'order': 4},\n",
      "                                                 {'begin': '15589344',\n",
      "                                                  'end': '15589456',\n",
      "                                                  'order': 5},\n",
      "                                                 {'begin': '15587753',\n",
      "                                                  'end': '15587858',\n",
      "                                                  'order': 6},\n",
      "                                                 {'begin': '15585475',\n",
      "                                                  'end': '15585572',\n",
      "                                                  'order': 7},\n",
      "                                                 {'begin': '15581221',\n",
      "                                                  'end': '15581390',\n",
      "                                                  'order': 8},\n",
      "                                                 {'begin': '15578089',\n",
      "                                                  'end': '15578315',\n",
      "                                                  'order': 9},\n",
      "                                                 {'begin': '15575666',\n",
      "                                                  'end': '15575810',\n",
      "                                                  'order': 10},\n",
      "                                                 {'begin': '15573367',\n",
      "                                                  'end': '15573465',\n",
      "                                                  'order': 11},\n",
      "                                                 {'begin': '15572201',\n",
      "                                                  'end': '15572323',\n",
      "                                                  'order': 12},\n",
      "                                                 {'begin': '15571624',\n",
      "                                                  'end': '15571796',\n",
      "                                                  'order': 13},\n",
      "                                                 {'begin': '15570295',\n",
      "                                                  'end': '15570353',\n",
      "                                                  'order': 14},\n",
      "                                                 {'begin': '15567726',\n",
      "                                                  'end': '15567826',\n",
      "                                                  'order': 15},\n",
      "                                                 {'begin': '15566253',\n",
      "                                                  'end': '15566369',\n",
      "                                                  'order': 16},\n",
      "                                                 {'begin': '15564024',\n",
      "                                                  'end': '15564218',\n",
      "                                                  'order': 17},\n",
      "                                                 {'begin': '15561033',\n",
      "                                                  'end': '15562013',\n",
      "                                                  'order': 18}]},\n",
      "                             'length': 3339,\n",
      "                             'name': 'transcript variant 1',\n",
      "                             'protein': {'accessionVersion': 'NP_001358344.1',\n",
      "                                         'ensemblProtein': 'ENSP00000252519.3',\n",
      "                                         'length': 805,\n",
      "                                         'name': 'angiotensin-converting '\n",
      "                                                 'enzyme 2 precursor'},\n",
      "                             'type': 'PROTEIN_CODING'}],\n",
      "            'type': 'PROTEIN_CODING'}]}\n"
     ]
    }
   ],
   "source": [
    "with open('sars-cov2-gene-data/ncbi_dataset/data/data_report.yaml') as yaml_file:\n",
    "    gene_data = yaml.load(yaml_file, Loader=yaml.SafeLoader)\n",
    "pprint(gene_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_strand(orientation):\n",
    "    \"\"\"Convert NCBI Dataset data report orientation to BioPython strand.\n",
    "    \"\"\"\n",
    "    return {'minus':-1, 'plus':+1, None:None}[orientation]\n",
    "\n",
    "def make_simple_location(interval, ref, default_strand=None):\n",
    "    \"\"\"Convert NCBI Dataset data report range element to BioPython\n",
    "    FeatureLocation.\n",
    "    \n",
    "    Note: Data report is 1-based, BioPython is 0-based.\n",
    "    \"\"\"\n",
    "    strand = get_orientation(interval.get('orientation', None))\n",
    "    if strand is None:\n",
    "        strand = default_strand\n",
    "    return FeatureLocation(int(interval['begin'])-1,\n",
    "                           int(interval['end'])-1,\n",
    "                           strand,\n",
    "                           ref)\n",
    "\n",
    "def make_location(location_data, default_strand=None):\n",
    "    \"\"\"Convert NCBI Dataset data report location to BioPython\n",
    "    FeatureLocation or CompoundLocation.\n",
    "    \n",
    "    Note: Data report is 1-based, BioPython is 0-based.\n",
    "    \"\"\"\n",
    "\n",
    "    ref = location_data['accessionVersion']\n",
    "    if len(location_data['range']) == 1:\n",
    "        return make_simple_location(location_data['range'][0],\n",
    "                                    ref,\n",
    "                                    default_strand)\n",
    "\n",
    "    loc = []\n",
    "    for interval in location_data['range']:\n",
    "        loc.append(make_simple_location(interval, ref, default_strand))\n",
    "    return CompoundLocation(loc)\n",
    "\n",
    "def clip_location(position):\n",
    "    \"\"\"Clip a postion for a BioPython FeatureLocation to the reference.\n",
    "    \n",
    "    WARNING: Use of reference length not implemented yet.\n",
    "    \"\"\"\n",
    "    return max(0, position)\n",
    "\n",
    "def gene_to_upstream(gene, upstream_length = 2000):\n",
    "    \"\"\"Given a gene structure from an NCBI Dataset gene data report,\n",
    "    return a list of rows with upstream range and the set of\n",
    "    transcripts sharing that upstream range.\n",
    "    \n",
    "    Example: [[FeatureLocation(...), 'NM_123', 'NM_456']]\n",
    "    \"\"\"\n",
    "    upstream_collection = dict()\n",
    "    # Need strand from gene location, since it's not included in exon locations.\n",
    "    gene_location = make_location(gene['genomicRanges'][0])\n",
    "\n",
    "    for transcript in gene['transcripts']:\n",
    "        transcript_accession = transcript['accessionVersion']\n",
    "        transcript_location = make_location(transcript['exons'], gene_location.strand)\n",
    "\n",
    "        if transcript_location.strand < 0:\n",
    "            upstream = FeatureLocation(max(transcript_location) + 1,\n",
    "                                       max(transcript_location) + upstream_length,\n",
    "                                       transcript_location.strand,\n",
    "                                       gene_location.ref)\n",
    "        else:\n",
    "            upstream = FeatureLocation(clip_location(min(transcript_location) - upstream_length),\n",
    "                                       clip_location(min(transcript_location) - 1),\n",
    "                                       transcript_location.strand,\n",
    "                                       gene_location.ref)\n",
    "\n",
    "        # BioSeq FeatureLocation not hashable.\n",
    "        upstream_key = (min(upstream), max(upstream), upstream.strand, upstream.ref)\n",
    "        if upstream_key in upstream_collection:\n",
    "            upstream_collection[upstream_key][1].append(transcript_accession)\n",
    "        else:\n",
    "            upstream_collection[upstream_key] = [upstream, transcript_accession]\n",
    "    return list(upstream_collection.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Report for gene ACE2\n",
      "[[FeatureLocation(ExactPosition(15602157), ExactPosition(15604156), strand=-1, ref='NC_000023.11'), 'NM_021804.3'], [FeatureLocation(ExactPosition(15600959), ExactPosition(15602958), strand=-1, ref='NC_000023.11'), 'NM_001371415.1']]\n"
     ]
    }
   ],
   "source": [
    "for gene in gene_data['genes']:\n",
    "    print(f'Report for gene {gene[\"symbol\"]}')\n",
    "    upstream_collection = gene_to_upstream(gene)\n",
    "    print(upstream_collection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
